### ISSUE TITLE
Program transformation to uncurry functions for which all applications are fully saturated.

### ISSUE THREAD

**Issue by Unisay** [Unisay/purescript-lua#24]
```lua
local add = function(x) 
  return function(y) 
    return x + y 
  end 
end
```
is two times slower than 

```lua
local add = function(x, y) 
  return x + y 
end
```

The idea is to rewrite former to latter but only for functions that are always fully-applied, e.g. a partial application
```lua
add(i)
```
never happens in the program being compiled;

The applications need to be rewritten too:
```lua
add(x)(y) ===> add(x,y)
```

**Comment by Unisay** [comment]
Upgrading this with a concrete design from the 2026-07 performance series (#172, #178, #179, #180).

**Drop the "never partially applied" condition.** The whole-program saturation analysis is unnecessary with a worker/wrapper split: every binding with manifest arity n ≥ 2 becomes a worker `f$w = AbsN [x₁…xₙ] body` (a real n-ary Lua function) plus a wrapper `f = λx₁.…λxₙ. AppN f$w [x₁…xₙ]` (curried, marked inline-always). Saturated sites become direct worker calls through wrapper inlining plus beta reduction (#180 provides the call-site inliner, #179 the `AppN` node); partial applications keep going through the wrapper. Semantics is unchanged by construction, and a mixed-use function wins at its saturated sites instead of being disqualified entirely.

**Eta-soundness.** Note [Eta reduction is unsound] is not disturbed: the worker body runs only on full saturation, so no work moves to construction time — the failure mode that killed eta reduction (#32) cannot arise from this scheme.

**`AbsN`.** The definition-side counterpart of #179's `AppN`, introduced the same way: `Abs` becomes the singleton pattern synonym (`pattern Abs ann p b = AbsN ann (p :| []) b`), so unary rules compile unchanged. `mkFnN` lands here too (deferred from #179 — its body is an n-ary function literal).

**Updated numbers.** The issue body says "two times slower"; measured on the current output it is 4.3x per application under PUC 5.1 — and under LuaJIT it is a cliff, not a constant: closure creation (`FNEW`) is NYI for the trace recorder, every partial application executes one, so every trace through a curried hot loop aborts and the loop start gets blacklisted. Measured: the curried loop runs 0.91 s with the JIT on and 0.90 s with `-joff` (the JIT contributes nothing), the uncurried equivalent 0.017 s (54x); driving the linked `Data.Array.foldl` with a curried step, 0.99 s vs 0.012 s (85x). Uncurrying is the main JIT unlock for the OpenResty target.

**Relations.** Prerequisite: #179 (`AppN`). Synergy: #180 (wrapper inlining), #178 (primops in worker bodies). Unblocks loopification of self-recursive tail calls (separate issue, will reference this one). Measured end-to-end by #172. Structural golden churn will be large but mechanical.

**Comment by Unisay** [comment]
A note on the saturation analysis, recorded so it does not get lost: the analysis will only see arity in manifest lambda chains. A definition like `handler = compose validate log` has semantic arity 1 but manifest arity 0, so it gets skipped even when every call site is saturated. Eta expanding it first (`handler = \ev -> compose validate log ev`) would expose the arity, but at the cost of sharing: the prefix `compose validate log` would then be recomputed on every call. The expansion is only safe when the prefix is cheap (a `Ref` or a literal, the same class `betaReduce`'s `isInlinableExpr` accepts). This is an extension, not a prerequisite: the conservative version that only uncurries manifest `Abs` chains is correct and already worth shipping.
